src/funcCoreTools/funcHostTask.ts

```diff
@@ -32,43 +53,58 @@ export function registerFuncHostTaskEvents(): void {
         context.errorHandling.suppressDisplay = true;
         context.telemetry.suppressIfSuccessful = true;
 
+        // Used to stop the task started with pickFuncProcess.ts startFuncProcessFromApi.
+        if (DotnetDebugDebugConfiguration.is(debugSession.configuration) && debugSession.configuration.launchServiceData.buildPath) {
+            const buildPathUri: vscode.Uri = vscode.Uri.file(debugSession.configuration.launchServiceData.buildPath)
+            stopFuncTaskIfRunning(buildPathUri.fsPath, /* terminate */ true)
+        }
+
         // NOTE: Only stop the func task if this is the root debug session (aka does not have a parentSession) to fix https://github.com/microsoft/vscode-azurefunctions/issues/2925
         if (getWorkspaceSetting<boolean>('stopFuncTaskPostDebug') && !debugSession.parentSession && debugSession.workspaceFolder) {
-            stopFuncTaskIfRunning(debugSession.workspaceFolder);
+            stopFuncTaskIfRunning(debugSession.workspaceFolder.uri.fsPath);
         }
     });
 }
 
-export function stopFuncTaskIfRunning(workspaceFolder: vscode.WorkspaceFolder): void {
-    const runningFuncTask: IRunningFuncTask | undefined = runningFuncTaskMap.get(workspaceFolder);
+export function stopFuncTaskIfRunning(functionApp: string, terminate: boolean = false): void {
+    const runningFuncTask: IRunningFuncTask | undefined = runningFuncTaskMap.get(functionApp);
     if (runningFuncTask !== undefined) {
-        // Use `process.kill` because `TaskExecution.terminate` closes the terminal pane and erases all output
-        // Also to hopefully fix https://github.com/microsoft/vscode-azurefunctions/issues/1401
-        process.kill(runningFuncTask.processId);
-        runningFuncTaskMap.delete(workspaceFolder);
+        if (terminate) {
+            // Tasks that are spun up by an API will execute quickly that process.kill does not terminate the func host fast enough
+            // that it hangs on to the port needed by a debug-restart event.
+            runningFuncTask.taskExecution.terminate();
+        }
+        else {
+            // Use `process.kill` because `TaskExecution.terminate` closes the terminal pane and erases all output
+            // Also to hopefully fix https://github.com/microsoft/vscode-azurefunctions/issues/1401
+            process.kill(runningFuncTask.processId);
+        }
+        runningFuncTaskMap.delete(functionApp);
     }
 }
 
-export async function getFuncPortFromTaskOrProject(context: IActionContext, funcTask: vscode.Task | undefined, projectPathOrTaskScope: string | vscode.WorkspaceFolder | vscode.TaskScope): Promise<string> {
+export async function getFuncPortFromTaskOrProject(context: IActionContext, funcTask: vscode.Task | undefined, projectPathOrTaskScope: string | vscode.WorkspaceFolder | vscode.TaskScope | undefined): Promise<string> {
```

Comment:
What does it mean for this value to be undefined now?

Reply:
There no `project` to search for the function's port and will rely on the task to get the running port.